File: ruby/213.house-robber-ii.rb
# @param {Integer[]} nums
# @return {Integer}
def rob(nums)
  return nums[0] if nums.size == 1

  rob_first = rob_linear(nums[...nums.size - 1])
  not_rob_first = rob_linear(nums[1..])

  [rob_first, not_rob_first].max
end

def rob_linear(nums)
  a, b = nums.reduce([0, 0]) do |(prev_a, prev_b), num|
    [prev_b + num, [prev_a, prev_b].max]
  end

  [a, b].max
end

File: rust/src/213.house-robber-ii.rs
use crate::Solution;

impl Solution {
    pub fn rob(nums: Vec<i32>) -> i32 {
        let length = nums.len();
        let first_value = nums[0];

        let (a, b) = nums.iter().enumerate().skip(1).fold(
            (first_value, first_value),
            |(prev_a, prev_b), (i, &num)| {
                if i == 1 || i == length - 1 {
                    (prev_a, prev_b)
                } else {
                    (prev_b + num, prev_a.max(prev_b))
                }
            },
        );

        let ans = a.max(b);

        let (a, b) = nums.iter().skip(1).fold((0, 0), |(prev_a, prev_b), &num| {
            (prev_b + num, prev_a.max(prev_b))
        });
        ans.max(a.max(b))
    }
}
